### PR TITLE
Pockets: Add 'Ad Free' functionality

### PR DESCRIPTION
**Pull request by beckyvb**
This PR adds functionality to hide pockets that are marked as 'ads' for any user with the Garden.AdFree.Allow permission.

Add 'Custom' pocket location
Add permission setting (Garden.AdFree.Allow) 
Add 'Ad' type to pockets
Validate Smarty-rendered pockets with 'Custom' locations (using CanRender function)
Use javascript to hide Repeat options when 'Custom' location is set.

### CODE REVIEW

**Comment by beckyvb** [comment]
Relies on https://github.com/vanilla/vanilla/pull/2558

File: plugins/Pockets/library/class.pocket.php [hunk 1]
```diff
@@ -53,9 +56,12 @@ class Pocket {
    /** $var array The repeat frequency. */
    public $MobileNever = FALSE;
 
+   /** $var boolean Is the pocket an ad? */
+   public $IsAd = FALSE;
```
Comment (tburry) [inline plugins/Pockets/library/class.pocket.php]: The property should be the type and there should be an `IsAd()` convenience method.

File: plugins/Pockets/class.pockets.plugin.php [hunk 2]
```diff
@@ -391,7 +398,14 @@ public static function PocketString($Name, $Data = NULL) {
 
       $Result = '';
       foreach ($Pockets as $Pocket) {
-         $Result .= $Pocket->ToString();
+         if (val('Location', $Pocket) == 'Custom' ) {
+            $Data['PageName'] = Pocket::PageName($Data['sender']->Controller->ControllerName);
```
Comment (tburry) [inline plugins/Pockets/class.pockets.plugin.php]: Use `Gdn::Controller()` and assign a variable before the loop.

File: plugins/Pockets/library/class.pocket.php [hunk 3]
```diff
@@ -22,6 +22,9 @@ class Pocket {
    const REPEAT_EVERY = 'every';
    const REPEAT_INDEX = 'index';
 
+   const AD_TYPE = 'ad';
```
Comment (tburry) [inline plugins/Pockets/library/class.pocket.php]: Constants are almost always in the form `TYPE_*` rather than `*_TYPE`.

**Pull request merged**